File: Lab2/Conversion/mm_C++/mastermind.cpp
#include "builtin.hpp"
#include "math.hpp"
#include "time.hpp"
#include "re.hpp"
#include "sys.hpp"
#include "random.hpp"
#include "code.hpp"
#include "board.hpp"
#include "mastermind.hpp"
#include "row.hpp"
#include "colour.hpp"
#include "peg.hpp"
#include "game.hpp"

namespace __mastermind__ {

str *const_0, *const_1, *const_10, *const_11, *const_12, *const_13, *const_14, *const_15, *const_16, *const_17, *const_18, *const_19, *const_2, *const_20, *const_21, *const_22, *const_23, *const_3, *const_4, *const_5, *const_6, *const_7, *const_8, *const_9;


str *__name__;



/**
 copyright Sean McCarthy, license GPL v2 or later 
*/
class_ *cl_Mastermind;

void *Mastermind::play(__ss_int guesses) {
    __game__::Game *gm;

    this->greeting();
    gm = (new __game__::Game(guesses));

    while (__NOT(gm->isOver())) {
        print2(NULL,0,4, const_0, ___box((gm->getTries()+1)), const_1, ___box(gm->getMaxTries()));
        gm->makeGuess(this->__readGuess());
        print2(NULL,0,1, const_2);
        this->display(gm);
        print2(NULL,0,1, const_3);
    }
    if (gm->isWon()) {
        print2(NULL,0,1, const_4);
    }
    else {
        print2(NULL,1,1, const_5);
        this->displaySecret(gm);
    }
    return NULL;
}

void *Mastermind::greeting() {
    
    print2(NULL,0,1, const_6);
    print2(NULL,0,1, const_3);
    print2(NULL,0,1, const_7);
    print2(NULL,0,1, const_3);
    print2(NULL,0,1, const_6);
    print2(NULL,0,1, const_8);
    print2(NULL,0,1, const_9);
    print2(NULL,0,1, const_6);
    return NULL;
}

__code__::Code *Mastermind::__readGuess() {
    list<str *> *__45, *colours;
    __peg__::Peg *peg;
    list<__peg__::Peg *> *guessPegs;
    __ss_bool inputOk;
    __iter<str *> *__46;
    list<str *>::for_in_loop __48;
    str *c;
    __ss_int __47;

    guessPegs = (new list<__peg__::Peg *>());
    print2(NULL,0,1, const_10);
    print2(NULL,0,1, const_11);
    inputOk = False;

    while (__NOT(inputOk)) {
        colours = __re__::split(const_12, raw_input(const_13), 4);

        FOR_IN(c,colours,45,47,48)
            peg = this->__parseColour(c);
            if ((peg!=NULL)) {
                guessPegs->append(peg);
            }
        END_FOR

        inputOk = ___bool((len(guessPegs)==4));
        if (__NOT(inputOk)) {
            print2(NULL,0,1, const_14);
            guessPegs = (new list<__peg__::Peg *>());
        }
    }
    return (new __code__::Code(guessPegs));
}

__peg__::Peg *Mastermind::__parseColour(str *s) {
    
    if ((__re__::search(const_15, s, 0)!=NULL)) {
        return (new __peg__::Peg(__colour__::Colours::red));
    }
    else if ((__re__::search(const_16, s, 0)!=NULL)) {
        return (new __peg__::Peg(__colour__::Colours::purple));
    }
    else if ((__re__::search(const_17, s, 0)!=NULL)) {
        return (new __peg__::Peg(__colour__::Colours::green));
    }
    else if ((__re__::search(const_18, s, 0)!=NULL)) {
        return (new __peg__::Peg(__colour__::Colours::yellow));
    }
    else if ((__re__::search(const_19, s, 0)!=NULL)) {
        return (new __peg__::Peg(__colour__::Colours::white));
    }
    else if ((__re__::search(const_20, s, 0)!=NULL)) {
        return (new __peg__::Peg(__colour__::Colours::black));
    }
    else {
        return NULL;
    }
    return 0;
}

void *Mastermind::display(__game__::Game *game) {
    list<__row__::Row *> *__21;
    list<__peg__::Peg *>::for_in_loop __30, __36;
    __board__::Board *__26;
    __code__::Code *__32, *__38;
    list<__row__::Row *>::for_in_loop __24;
    list<__peg__::Peg *> *__27, *__33;
    void *__25, *__31, *__37;
    __iter<__peg__::Peg *> *__28, *__34;
    __peg__::Peg *p;
    __ss_int __23, __29, __35;
    __row__::Row *r;
    __iter<__row__::Row *> *__22;


    FOR_IN(r,(game->getBoard())->getRows(),21,23,24)

        FOR_IN(p,(r->getGuess())->getPegs(),27,29,30)
            print2(NULL,1,1, (__str(__colour__::getColourName(p->getColour())))->rjust(6));
        END_FOR

        print2(NULL,1,1, const_21);

        FOR_IN(p,(r->getResult())->getPegs(),33,35,36)
            print2(NULL,1,1, (__str(__colour__::getColourName(p->getColour())))->rjust(6));
        END_FOR

        print2(NULL,0,1, const_6);
    END_FOR

    return NULL;
}

void *Mastermind::displaySecret(__game__::Game *game) {
    list<__peg__::Peg *>::for_in_loop __42;
    list<__peg__::Peg *> *__39;
    __code__::Code *__44;
    void *__43;
    __iter<__peg__::Peg *> *__40;
    __peg__::Peg *p;
    __ss_int __41;


    FOR_IN(p,(game->getSecretCode())->getPegs(),39,41,42)
        print2(NULL,1,1, (__str(__colour__::getColourName(p->getColour())))->rjust(6));
    END_FOR

    return NULL;
}

/**
Instantiate mastermind and invoke play method to play game

*/
void *__ss_main() {
    Mastermind *m;
    __ss_bool __49, __50;
    __ss_int guesses;

    m = (new Mastermind());
    guesses = 8;
    if (((len(__sys__::argv)>1) and (__re__::match(const_22, (__sys__::argv)->__getfast__(1), 0)!=NULL))) {
        guesses = __int((__sys__::argv)->__getfast__(1));
    }
    m->play(guesses);
    return NULL;
}

void __init() {
    const_0 = new str("Guess: ");
    const_1 = __char_cache[47];;
    const_2 = new str("--------Board--------");
    const_3 = new str("---------------------");
    const_4 = new str("Congratulations!");
    const_5 = new str("Secret Code: ");
    const_6 = new str("");
    const_7 = new str("Welcome to Mastermind");
    const_8 = new str("Each guess should be 4 colours from any of:");
    const_9 = new str("red yellow green purple black white");
    const_10 = new str("Type four (space seperated) colours from:");
    const_11 = new str("[r]ed [y]ellow [g]reen [p]urple [b]lack [w]hite");
    const_12 = new str("\\s");
    const_13 = new str("> ");
    const_14 = new str("Invalid input, use colours as stated");
    const_15 = new str("^r");
    const_16 = new str("^p");
    const_17 = new str("^g");
    const_18 = new str("^y");
    const_19 = new str("^w");
    const_20 = new str("^b");
    const_21 = new str(" | Result:\t");
    const_22 = new str("\\d");
    const_23 = new str("__main__");

    __name__ = new str("__main__");

    /**
     copyright Sean McCarthy, license GPL v2 or later 
    */
    cl_Mastermind = new class_("Mastermind");
    /**
    Instantiate mastermind and invoke play method to play game
    
    */
    if (__eq(__name__, const_23)) {
        __ss_main();
    }
}

} // module namespace

int main(int __ss_argc, char **__ss_argv) {
    __shedskin__::__init();
    __math__::__init();
    __time__::__init();
    __random__::__init();
    __colour__::__init();
    __peg__::__init();
    __code__::__init();
    __row__::__init();
    __board__::__init();
    __game__::__init();
    __re__::__init();
    __sys__::__init(__ss_argc, __ss_argv);
    __shedskin__::__start(__mastermind__::__init);
}

File: Lab2/Conversion/mm_cpp2java_base/src/mastermind/__mastermind__/Mastermind.java
package mastermind.__mastermind__;

import java.io.BufferedInputStream;
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Iterator;

import mastermind.__board__.Board;
import mastermind.__code__.Code;
import mastermind.__colour__.Colours;
import mastermind.__game__.Game;
import mastermind.__peg__.Peg;
import mastermind.__row__.Row;

//C++ TO JAVA CONVERTER NOTE: Java has no need of forward class declarations:
//class Code;
//C++ TO JAVA CONVERTER NOTE: Java has no need of forward class declarations:
//class Board;
//C++ TO JAVA CONVERTER NOTE: Java has no need of forward class declarations:
//class Peg;
//C++ TO JAVA CONVERTER NOTE: Java has no need of forward class declarations:
//class Game;
//C++ TO JAVA CONVERTER NOTE: Java has no need of forward class declarations:
//class Row;
public class Mastermind {
/**
The game of Mastermind

This class provides a function "play" for invoking a new game.

The objective of the game is to guess a code composed of 4 coloured
pegs. The code can be composed of any combination of the six colours
(red, green, purple, yellow, white, black), and can include duplicates.

For each guess a result code may be returned composed of black and/or
white pegs. A black peg indicates a peg of the right colour and in the
right position. A white peg indicates a peg of the right colour but in
the wrong position. The arrangement of the pegs does not correspond to
the pegs in the guess- black pegs will always be shown first, followed
but white pegs.

The game ends with either a correct guess or after running out of 
guesses.
*/

	public static String const_0;
	public static String const_1;
	public static String const_10;
	public static String const_11;
	public static String const_12;
	public static String const_13;
	public static String const_14;
	public static String const_15;
	public static String const_16;
	public static String const_17;
	public static String const_18;
	public static String const_19;
	public static String const_2;
	public static String const_20;
	public static String const_21;
	public static String const_22;
	public static String const_23;
	public static String const_3;
	public static String const_4;
	public static String const_5;
	public static String const_6;
	public static String const_7;
	public static String const_8;
	public static String const_9;
	
	public static void __init() {
		const_0 = new String("Guess: ");
		const_1 = "";
		const_2 = new String("--------Board--------");
		const_3 = new String("---------------------");
		const_4 = new String("Congratulations!");
		const_5 = new String("Secret Code: ");
		const_6 = new String("");
		const_7 = new String("Welcome to Mastermind");
		const_8 = new String("Each guess should be 4 colours from any of:");
		const_9 = new String("red yellow green purple black white");
		const_10 = new String("Type four (space seperated) colours from:");
		const_11 = new String("[r]ed [y]ellow [g]reen [p]urple [b]lack [w]hite");
		const_12 = new String("\\s");
		const_13 = new String("> ");
		const_14 = new String("Invalid input, use colours as stated");
		const_15 = new String("^r");
		const_16 = new String("^p");
		const_17 = new String("^g");
		const_18 = new String("^y");
		const_19 = new String("^w");
		const_20 = new String("^b");
		const_21 = new String(" | Result:\t");
		const_22 = new String("\\d");
		const_23 = new String("__main__");

	}
	
	public static void main(String[] args) throws Exception {
		Main(args);
	}
	public static int Main( String[] __ss_argv) throws Exception {

		Colours.__init();
		
		Mastermind m;
		boolean __49;
		boolean __50;
		int guesses = 0;
		Mastermind.__init();

		 
		m = (new Mastermind());
		guesses = 8;
		if ((__ss_argv.length > 1) && (const_22.equals(__ss_argv[1]))) {
			guesses = Integer.parseInt(__ss_argv[1]);
		}
		m.play(guesses);
		
		return 0;
	}
	
	/**
	 * default constructor
	 */
	public Mastermind() {
	}
	
	/**
	 * Demarrage d'une partie
	 * @param guesses : nombre d'essais max
	 * @return 
	 */
	public final Object play(int guesses) throws Exception {
		Game gm;

		this.greeting();
		gm = (new Game(guesses));

		while (!gm.isOver()) {
			System.out.println(const_0 +"   " + (gm.getTries() + 1) +"   " + const_1 + "    " + gm.getMaxTries());
			gm.makeGuess(this.__readGuess());
			System.out.println(const_2);
			this.display(gm);
			System.out.println(const_3);
		}
		if (gm.isWon()) {
			System.out.println(const_4);
		}
		else {
			System.out.println(const_5);
			this.displaySecret(gm);
		}
		return null;
	}
	
	/**
	 * Affichage du msg de bienvenue
	 * @return 
	 */
	public final Object greeting() {

		System.out.println(const_6);
		System.out.println(const_3);
		System.out.println(const_7);
		System.out.println(const_3);
		System.out.println(const_6);
		System.out.println(const_8);
		System.out.println(const_9);
		System.out.println(const_6);
		return null;
	}
	
	/**
	 * Lecture de l'entree clavier pour avoir un 'code devinez'
	 * @return Code
	 */
	public final Code __readGuess() throws IOException {
		String[] __45;
		String[] colours;
		Peg peg;
		java.util.LinkedList<Peg> guessPegs;
		boolean inputOk = false;
		Iterator<String> __46;
		Iterator<String> __48;
		String c;
		int __47 = 0;

		guessPegs = (new java.util.LinkedList<Peg>());
		System.out.println(const_10);
		System.out.println(const_11);
		inputOk = false;

		while (!inputOk) {
			System.out.print(const_13);
			BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
			colours = br.readLine().split(const_12);
			
			__45 = colours;
			__47 = -1;
			//__48 = __45;
			for (int i = 0; i < colours.length; i++) {
				__47++;
				c = colours[i];
				System.out.println(c);
				peg = this.__parseColour(c);
				if ((peg != null)) {
					guessPegs.add(peg);
				}
			}

			inputOk = guessPegs.size() == 4;
			if (!inputOk) {
				System.out.println(const_14);
				guessPegs = (new java.util.LinkedList<Peg>());
			}
		}
		return (new Code(guessPegs));
	}
	
	/**
	 * Parcour de l'entree clavier (string) obtenue pour extraire les couleurs
	 * @param s : string du clavier
	 * @return Pion identifie
	 */
	public final Peg __parseColour(String s) {

		if (s.matches(const_15)) {
			return (new Peg(Colours.red));
		}
		else if (s.matches(const_16)) {
			return (new Peg(Colours.purple));
		}
		else if (s.matches(const_17)) {
			return (new Peg(Colours.green));
		}
		else if (s.matches(const_18)) {
			return (new Peg(Colours.yellow));
		}
		else if (s.matches(const_19)) {
			return (new Peg(Colours.white));
		}
		else if (s.matches(const_20)) {
			return (new Peg(Colours.black));
		}
		else {
			return null;
		}
	}
	
	/**
	 * Affiche de l'evolution du jeux
	 * @param game
	 * @return 
	 */

	public final Object display(Game game) {
		java.util.LinkedList<Row > __21;
		Iterator<Peg> __30;
		Iterator<Peg> __36;
		Board __26;
		Code __32;
		Code __38;
		Iterator<Row> __24;
		java.util.LinkedList<Peg > __27;
		java.util.LinkedList<Peg > __33;
		Object __25;
		Object __31;
		Object __37;
		Iterator<Peg > __28;
		Iterator<Peg > __34;
		Peg p;
		int __23 = 0;
		int __29 = 0;
		int __35 = 0;
		Row r;
		Iterator<Row > __22;


		__21 = (game.getBoard()).getRows();
		__23 = -1;
		__24 = __21.iterator();
		while (__24.hasNext()) {
			__23++;
			r = __24.next();

			__27 = (r.getGuess()).getPegs();
			__29 = -1;
			__30 = __27.iterator();
			while (__30.hasNext()) {
				__29++;
				p = __30.next();
				System.out.print(rjust(Colours.getColourName(p.getColour()),6));
			}

			System.out.print(const_21);

			__33 = (r.getResult()).getPegs();
			__35 = -1;
			__36 = __33.iterator();
			while (__36.hasNext()) {
				__35++;
				p = __36.next();
				System.out.print(rjust(Colours.getColourName(p.getColour()),6));
			}

			System.out.println(const_6);
		}

		return null;
	}
	
	/**
   	 * Affichage d'un pion en remplissant d'espace 
   	 * pour uniformiser la taille
   	 */
	public static String rjust(String s, int len){
		int space = len+1-s.length(), i;
    	for(i=0; space>i; i++){ //.rjust(6)
    		s = " " + s;
    	}
    	return s;
	}
	
	/**
	 * Affichage du code secret
	 * @param game
	 * @return 
	 */
	public final Object displaySecret(Game game) {
		Iterator<Peg> __42;
		java.util.LinkedList<Peg > __39;
		Code __44;
		Object __43;
		Iterator<Peg > __40;
		Peg p;
		int __41 = 0;


		__39 = (game.getSecretCode()).getPegs();
		__41 = -1;
		__42 = __39.iterator();
		while (__42.hasNext()) {
			__41++;
			p = __42.next();
			System.out.print(rjust(Colours.getColourName(p.getColour()),6));
		}

		return null;
	}
}
//C++ TO JAVA CONVERTER TODO TASK: Only the namespaces at the beginning of the file can be converted to the Java 'package' for this file:
//ORIGINAL LINE: namespace __mastermind__
